server/com.dexels.navajo.compiler.tsl/src/com/dexels/navajo/compiler/tsl/internal/BundleQueueComponent.java

```diff
@@ -170,18 +173,39 @@ private void checkForChangedScripts(Event e, String folder) {
                     	continue;
                     }
                     enqueueScript(scriptName, changedScript);
-                    enqueueDependentScripts(scriptName, new HashSet<String>());
                 }
             } catch (IllegalArgumentException e1) {
                 logger.warn("Error in handling changed script {}: {}", changedScript, e1);
             }
         }
     }
 
+    // check whether dependencies of the current script are satisfied, and if not create them on demand
+    private void checkOnDemandScriptDependencies(String script) {
+        String rpcName = script;
+        String bareScript = script.substring(script.lastIndexOf('/') + 1);
+        if (bareScript.indexOf('_') >= 0) {
+            rpcName = script.substring(0, script.lastIndexOf('_'));
+        }
 
-    private void enqueueDependentScripts(String script, Set<String> history) {
-        history.add(script);
-        
+        // For now, only entity dependencies are relevant script dependencies.
+        // For instance, in the case the server runs in DEVELOP_MODE where
+        // entities are lazily loaded, all bundles for super entities also need
+        // to be installed
+        List<Dependency> dependencies = depanalyzer.getDependencies(rpcName, Dependency.ENTITY_DEPENDENCY);
+
+        for (Dependency dependency : dependencies) {
+            String depScript = dependency.getDependee();
+            try {
+                bundleCreator.getOnDemandScriptService(depScript, null);
+                checkOnDemandScriptDependencies(depScript);
+            } catch (CompilationException e) {
+                logger.info("Failed to recompile {} after a change in {}: {}", depScript, script, e);
+            }
```

Comment:
Recursion. Is anything done to prevent recursion or has this been inspected to be absolutely certain that it cannot go into infinite loop?

Reply:
Why are we not using enqueueScript here, similarly to enqueueDependentScripts? And let that handle the recursion.